File: 37.sudoku-solver.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=37 lang=java
 *
 * [37] Sudoku Solver
 */

// @lc code=start

class Solution {
    //
    private boolean[][] row = new boolean[9][9];
    private boolean[][] column = new boolean[9][9];
    private boolean[][][] block = new boolean[3][3][9];
    private boolean valid = false;
    // the space of empty units
    private List<int[]> spaces = new ArrayList<int[]>();

    public void solveSudoku(char[][] board) {
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.') {
                    spaces.add(new int[] { i, j });
                } else {
                    // the range is [1,9], but idx starts from 0
                    int digit = board[i][j] - '0' - 1;
                    // digit is in row i, column j, and block[i/3][j/3]
                    row[i][digit] = column[j][digit] = block[i / 3][j / 3][digit] = true;
                }
            }
        }

        dfs(board, 0);
    }

    // cnt is the no. of filled units which were empty
    public void dfs(char[][] board, int cnt) {
        if (cnt == spaces.size()) {
            valid = true;
            return;
        }

        int[] space = spaces.get(cnt);
        int i = space[0], j = space[1];
        for (int digit = 0; digit < 9 && !valid; ++digit) {
            if (!row[i][digit] && !column[j][digit] && !block[i / 3][j / 3][digit]) {
                row[i][digit] = column[j][digit] = block[i / 3][j / 3][digit] = true;
                board[i][j] = (char) (digit + '0' + 1);
                dfs(board, cnt + 1);
                row[i][digit] = column[j][digit] = block[i / 3][j / 3][digit] = false;
            }
        }
    }
}

// @lc code=end

File: 37.sudoku-solver.py
#
# @lc app=leetcode id=37 lang=python3
#
# [37] Sudoku Solver
#

# @lc code=start
from typing import List


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Solve Sudoku in-place using backtracking + bitmasks.
        The board is a 9x9 grid with digits '1'..'9' and '.' for empty cells.
        """

        # row_mask[i], col_mask[j], box_mask[b] are 9-bit masks indicating
        # which digits (1..9) are already used in row i / col j / 3x3 box b.
        # Bit k (0-based) corresponds to digit (k+1).
        row_mask = [0] * 9
        col_mask = [0] * 9
        box_mask = [0] * 9

        empties = []  # list of coordinates (i, j) for empty cells

        def box_id(i: int, j: int) -> int:
            """Return 0..8 index of 3x3 sub-box containing (i, j)."""
            return (i // 3) * 3 + (j // 3)

        def set_digit(i: int, j: int, d: int) -> None:
            """Place digit d (0-based: 0..8 => '1'..'9') at (i, j) and set masks."""
            bit = 1 << d
            board[i][j] = chr(ord("1") + d)
            row_mask[i] |= bit
            col_mask[j] |= bit
            box_mask[box_id(i, j)] |= bit

        def unset_digit(i: int, j: int, d: int) -> None:
            """Remove digit d from (i, j) and unset masks (backtracking)."""
            bit = 1 << d
            board[i][j] = "."
            row_mask[i] ^= bit
            col_mask[j] ^= bit
            box_mask[box_id(i, j)] ^= bit

        # 1) Initialize masks from the given board and collect empty cells.
        for i in range(9):
            for j in range(9):
                c = board[i][j]
                if c == ".":
                    empties.append((i, j))
                else:
                    d = ord(c) - ord("1")  # 0..8
                    set_digit(i, j, d)  # sets masks only (board stays same)

        # 2) Backtracking: fill empties[idx:] one by one.
        #    We choose the next cell greedily (MRV heuristic) to speed up:
        #    pick the empty with the fewest candidates.
        def solve(idx: int) -> bool:
            if idx == len(empties):
                return True  # all filled

            # Pick the best position (minimum remaining candidates)
            best = idx
            best_count = 10  # larger than any possible candidate count (<=9)
            for k in range(idx, len(empties)):
                i, j = empties[k]
                used = row_mask[i] | col_mask[j] | box_mask[box_id(i, j)]
                cand = (~used) & 0x1FF  # available digits (9 bits)
                cnt = cand.bit_count()
                if cnt < best_count:
                    best_count = cnt
                    best = k
                    if cnt == 1:
                        break  # can't get better than 1

            # If no candidate exists for some cell, prune
            if best_count == 0:
                return False

            # Swap the chosen cell to position idx
            empties[idx], empties[best] = empties[best], empties[idx]
            i, j = empties[idx]

            # Try all candidates for (i, j)
            used = row_mask[i] | col_mask[j] | box_mask[box_id(i, j)]
            cand = (~used) & 0x1FF  # 9-bit mask of candidates

            # Iterate set bits of cand using lowbit trick
            while cand:
                low = cand & -cand  # lowest set bit
                d = low.bit_length() - 1  # 0..8 digit index
                set_digit(i, j, d)
                if solve(idx + 1):
                    return True
                unset_digit(i, j, d)
                cand ^= low  # remove this candidate and continue

            # Restore the order (optional for correctness but neat)
            empties[idx], empties[best] = empties[best], empties[idx]
            return False

        solve(0)


# @lc code=end
